source/distributed/error_predictor.cc

```diff
@@ -0,0 +1,360 @@
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    ErrorPredictor<dim, spacedim>::prepare_for_coarsening_and_refinement(
+      const std::vector<const Vector<float> *> &all_in,
+      const double                              gamma_p,
+      const double                              gamma_h,
+      const double                              gamma_n)
+    {
+      error_indicators = all_in;
+      this->gamma_p    = gamma_p;
+      this->gamma_h    = gamma_h;
+      this->gamma_n    = gamma_n;
+      register_data_attach();
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    ErrorPredictor<dim, spacedim>::register_data_attach()
+    {
+      // TODO: casting away constness is bad
+      parallel::distributed::Triangulation<dim, spacedim> *tria =
+        (dynamic_cast<parallel::distributed::Triangulation<dim, spacedim> *>(
+          const_cast<dealii::Triangulation<dim, spacedim> *>(
+            &dof_handler->get_triangulation())));
+      Assert(tria != nullptr, ExcInternalError());
+
+      handle = tria->register_data_attach(
+        [this](const typename Triangulation<dim, spacedim>::cell_iterator &cell,
+               const typename Triangulation<dim, spacedim>::CellStatus status) {
+          return this->pack_callback(cell, status);
+        },
+        /*returns_variable_size_data=*/false);
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    ErrorPredictor<dim, spacedim>::prepare_for_coarsening_and_refinement(
+      const Vector<float> &in,
+      const double         gamma_p,
+      const double         gamma_h,
+      const double         gamma_n)
+    {
+      std::vector<const Vector<float> *> all_in(1, &in);
+      prepare_for_coarsening_and_refinement(all_in, gamma_p, gamma_h, gamma_n);
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    ErrorPredictor<dim, spacedim>::unpack(std::vector<Vector<float> *> &all_out)
+    {
+      Assert(error_indicators.size() == all_out.size(),
+             ExcDimensionMismatch(error_indicators.size(), all_out.size()));
+
+      // TODO: casting away constness is bad
+      parallel::distributed::Triangulation<dim, spacedim> *tria =
+        (dynamic_cast<parallel::distributed::Triangulation<dim, spacedim> *>(
+          const_cast<dealii::Triangulation<dim, spacedim> *>(
+            &dof_handler->get_triangulation())));
+      Assert(tria != nullptr, ExcInternalError());
+
+      tria->notify_ready_to_unpack(
+        handle,
+        [this, &all_out](
+          const typename Triangulation<dim, spacedim>::cell_iterator &cell,
+          const typename Triangulation<dim, spacedim>::CellStatus     status,
+          const boost::iterator_range<std::vector<char>::const_iterator>
+            &data_range) {
+          this->unpack_callback(cell, status, data_range, all_out);
+        });
+
+      for (const auto &out : all_out)
+        out->compress(::dealii::VectorOperation::insert);
```

Comment:
Is the `const` flag correct here?
`all_out` is of type `std::vector<Vector<float> *>`, so I guess a const reference to an immutable pointer should be okay.

Reply:
If it compiles, it is ok :-)